File: 025_reverse_nodes_in_k_group/nonrecursion/Solution.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        dummy_head = ListNode(-1)
        dummy_head.next = head
        pre, end = dummy_head, dummy_head

        while end.next:
            for i in range(k):
                end = end.next
                if not end: return dummy_head.next
            start = pre.next
            next_node = end.next
            end.next = None
            pre.next = self.reverse(start)
            start.next = next_node
            pre = start
            end = pre

        return dummy_head.next

    def reverse(self, head: ListNode):
        pre = None
        while head:
            next_node = head.next
            head.next, pre, head = pre, head, next_node

        return pre


if __name__ == '__main__':
    head = ListNode(1)
    cur = head
    for i in range(2, 10):
        cur.next = ListNode(i)
        cur = cur.next

    solution = Solution()
    cur = solution.reverseKGroup(head, 2)
    while cur:
        print(cur.val, end = ' ')
        cur = cur.next

File: 025_reverse_nodes_in_k_group/nonrecursion/Solution.go
package main

import "fmt"

type ListNode struct {
    Val  int
    Next *ListNode
}

func reverseKGroup(head *ListNode, k int) *ListNode {
    dummyHead := &ListNode{-1, head}
    pre, end := dummyHead, dummyHead

    for (end.Next != nil) {
        for i := 0; i < k; i++ {
            end = end.Next;
            if end == nil {
                return dummyHead.Next
            }
        }
        start := pre.Next
        next := end.Next
        end.Next = nil
        pre.Next = reverse(start)
        start.Next = next
        pre = start
        end = pre
    }

    return dummyHead.Next;
}

func reverse(head *ListNode) *ListNode {
    var pre *ListNode

    for head != nil {
        next := head.Next;
        head.Next, pre, head = pre, head, next
    }

    return pre
}

func printList(head *ListNode) {
    for head != nil {
        fmt.Printf("%d ", head.Val)
        head = head.Next
    }
    fmt.Println()
}

func main() {
    head := &ListNode{1, nil}
    cur := head
    for i := 2; i < 10; i++ {
        cur.Next = &ListNode{i, nil}
        cur = cur.Next
    }

    cur = reverseKGroup(head, 2)
    for cur != nil {
        fmt.Printf("%d ", cur.Val)
        cur = cur.Next
    }
}
